File: frontend/src/features/space-management/domain/constraintResolver.ts
// Port of backend/src/Dixels.Domain/SpaceManagement/ConstraintResolver.cs
//
// Stateless resolution of the Building -> Floor -> Space constraint hierarchy: a
// COALESCE-equivalent, per-field pick (space wins, then floor, then building), with
// provenance. The backend resolves against Building/Floor/Space entity classes; the
// frontend has no such entities yet (only mock data), so this operates on plain DTOs
// with the same field-precedence semantics instead.

import { OperatingDays } from './operatingDays'
import { OperatingWindow } from './operatingWindow'

export type ConstraintSource = 'Building' | 'Floor' | 'Space'

export interface FieldValue<T> {
  value: T
  source: ConstraintSource
}

export interface ResolvedConstraints {
  timezone: string
  days: FieldValue<OperatingDays>
  hours: FieldValue<OperatingWindow>
  maxDurationMinutes: FieldValue<number>
  maxHorizonDays: number
  minLeadMinutes: number
  minAttendees?: number
  capacity?: number
}

export interface BuildingConstraintInput {
  timezone: string
  days: OperatingDays
  hours: OperatingWindow
  maxDurationMinutes: number
  maxHorizonDays: number
  minLeadMinutes: number
}

export interface FloorConstraintInput {
  days?: OperatingDays
  hours?: OperatingWindow
  maxDurationMinutes?: number
}

export interface SpaceConstraintInput extends FloorConstraintInput {
  minAttendees?: number
  capacity?: number
}

export function resolve(
  building: BuildingConstraintInput,
  floor: FloorConstraintInput,
  space?: SpaceConstraintInput,
): ResolvedConstraints {
  return {
    timezone: building.timezone,
    days: pick(space?.days, floor.days, building.days),
    hours: pick(space?.hours, floor.hours, building.hours),
    maxDurationMinutes: pick(space?.maxDurationMinutes, floor.maxDurationMinutes, building.maxDurationMinutes),
    maxHorizonDays: building.maxHorizonDays,
    minLeadMinutes: building.minLeadMinutes,
    minAttendees: space?.minAttendees,
    capacity: space?.capacity,
  }
}

function pick<T>(spaceValue: T | undefined, floorValue: T | undefined, buildingValue: T): FieldValue<T> {
  if (spaceValue !== undefined) return { value: spaceValue, source: 'Space' }
  if (floorValue !== undefined) return { value: floorValue, source: 'Floor' }
  return { value: buildingValue, source: 'Building' }
}

export const DixelsDomainErrorCodes = {
  DaysNotNarrower: 'Dixels:SpaceManagement:DaysNotNarrower',
  HoursNotNarrower: 'Dixels:SpaceManagement:HoursNotNarrower',
} as const

export class ConstraintNarrowingError extends Error {
  readonly code: string
  constructor(code: string) {
    super(code)
    this.code = code
  }
}

export function ensureDaysNarrowing(childOwn: OperatingDays | undefined, resolvedParent: OperatingDays): void {
  if (childOwn !== undefined && !childOwn.isSubsetOf(resolvedParent)) {
    throw new ConstraintNarrowingError(DixelsDomainErrorCodes.DaysNotNarrower)
  }
}

export function ensureHoursNarrowing(childOwn: OperatingWindow | undefined, resolvedParent: OperatingWindow): void {
  if (childOwn !== undefined && !childOwn.isSubsetOf(resolvedParent)) {
    throw new ConstraintNarrowingError(DixelsDomainErrorCodes.HoursNotNarrower)
  }
}

export interface NarrowingCandidate {
  displayName: string
  ownDays?: OperatingDays
  ownHours?: OperatingWindow
}

/**
 * Given the descendants that have their own days/hours override set, returns the
 * display names of the ones that would no longer fit under a proposed, tighter parent
 * value. Non-blocking and informational — the save that tightens the parent still
 * proceeds (tightening never retroactively invalidates), but the admin sees exactly
 * which descendants now have a stale override.
 */
export function findNarrowingConflicts(
  candidates: NarrowingCandidate[],
  proposedDays: OperatingDays | undefined,
  proposedHours: OperatingWindow | undefined,
): string[] {
  const conflicts: string[] = []

  for (const candidate of candidates) {
    if (proposedDays !== undefined && candidate.ownDays !== undefined && !candidate.ownDays.isSubsetOf(proposedDays)) {
      conflicts.push(`${candidate.displayName} no longer fits inside the new operating days`)
    }

    if (proposedHours !== undefined && candidate.ownHours !== undefined && !candidate.ownHours.isSubsetOf(proposedHours)) {
      conflicts.push(`${candidate.displayName} no longer fits inside the new operating hours`)
    }
  }

  return conflicts
}

File: backend/src/Dixels.Domain/SpaceManagement/ConstraintResolver.cs
using System;
using System.Collections.Generic;
using Volo.Abp;
using Volo.Abp.Domain.Services;
using Dixels.SpaceManagement.ValueObjects;

namespace Dixels.SpaceManagement;

/// <summary>
/// Stateless resolution of the Building → Floor → Space constraint hierarchy: a
/// COALESCE-equivalent, per-field pick (space wins, then floor, then building), with
/// provenance. This is the single implementation used both when reading (the constraints
/// page's resolved-values strip) and when writing (narrowing validation on save).
/// </summary>
public class ConstraintResolver : IDomainService
{
    public ResolvedConstraints Resolve(Building building, Floor floor, Space? space = null)
    {
        var days = Pick(
            (space?.Days, ConstraintSource.Space),
            (floor.Days, ConstraintSource.Floor),
            (building.Days, ConstraintSource.Building));
        var hours = Pick(
            (space?.Hours, ConstraintSource.Space),
            (floor.Hours, ConstraintSource.Floor),
            (building.Hours, ConstraintSource.Building));
        var maxDurationMinutes = Pick(
            (space?.MaxDurationMinutes, ConstraintSource.Space),
            (floor.MaxDurationMinutes, ConstraintSource.Floor),
            (building.MaxDurationMinutes, ConstraintSource.Building));

        return new ResolvedConstraints(
            building.Timezone,
            days,
            hours,
            maxDurationMinutes,
            building.MaxHorizonDays,
            building.MinLeadMinutes,
            space?.MinAttendees,
            space?.Capacity);
    }

    public void EnsureDaysNarrowing(OperatingDays? childOwn, OperatingDays resolvedParent)
    {
        if (childOwn is not null && !childOwn.IsSubsetOf(resolvedParent))
        {
            throw new BusinessException(DixelsDomainErrorCodes.DaysNotNarrower);
        }
    }

    public void EnsureHoursNarrowing(OperatingWindow? childOwn, OperatingWindow resolvedParent)
    {
        if (childOwn is not null && !childOwn.IsSubsetOf(resolvedParent))
        {
            throw new BusinessException(DixelsDomainErrorCodes.HoursNotNarrower);
        }
    }

    /// <summary>
    /// Given the descendants that have their own days/hours override set, returns the
    /// display names of the ones that would no longer fit under a proposed, tighter parent
    /// value. This is a non-blocking, informational check — the save that tightens the
    /// parent still proceeds (tightening never retroactively invalidates), but the admin
    /// sees exactly which descendants now have a stale override.
    /// </summary>
    public IReadOnlyList<string> FindNarrowingConflicts(
        IEnumerable<NarrowingCandidate> candidates,
        OperatingDays? proposedDays,
        OperatingWindow? proposedHours)
    {
        var conflicts = new List<string>();

        foreach (var candidate in candidates)
        {
            if (proposedDays is not null && candidate.OwnDays is not null && !candidate.OwnDays.IsSubsetOf(proposedDays))
            {
                conflicts.Add($"{candidate.DisplayName} no longer fits inside the new operating days");
            }

            if (proposedHours is not null && candidate.OwnHours is not null && !candidate.OwnHours.IsSubsetOf(proposedHours))
            {
                conflicts.Add($"{candidate.DisplayName} no longer fits inside the new operating hours");
            }
        }

        return conflicts;
    }

    // Levels are walked in the order given — the first non-null value wins. Adding a level to
    // the hierarchy (e.g. a "Wing" between Building and Floor) means adding one more tuple at
    // each call site above; this method itself never needs to change (Open/Closed).
    private static FieldValue<T> Pick<T>(params (T? Value, ConstraintSource Source)[] levels)
        where T : class
    {
        foreach (var (value, source) in levels)
        {
            if (value is not null)
            {
                return new FieldValue<T>(value, source);
            }
        }

        throw new InvalidOperationException("The last level in the chain must always supply a non-null value.");
    }

    // Separate overload for value-typed fields (int) — C# can't express "T? that's either a
    // nullable reference or a Nullable<T> value type" with one unconstrained generic method, so
    // this duplication is a language limitation, not a hierarchy-precedence rule living in two
    // places.
    private static FieldValue<int> Pick(params (int? Value, ConstraintSource Source)[] levels)
    {
        foreach (var (value, source) in levels)
        {
            if (value.HasValue)
            {
                return new FieldValue<int>(value.Value, source);
            }
        }

        throw new InvalidOperationException("The last level in the chain must always supply a non-null value.");
    }
}
